drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1073 @@
+ * that needs to be fast.
+ * 
+ * There are three types of planes: primary, cursor and overlay. For
+ * compatibility with legacy userspace, the default behavior is to expose
+ * only overlay planes to userspace (we're going to see in the code that
+ * we have to ask to receive all types of planes). A good example of plane
+ * usage is this: imagine a static desktop screen and the user is moving
+ * the cursor around. Only the cursor is moving. Instead of calculating the
+ * complete scene for each time the user moves its cursor, we can update only
+ * the cursor plane and it will be automatically overlayed by the hardware on
+ * top of the primary plane. There's no need of software composition in this
+ * case.
+ *
+ * But there was synchronisation problems related to multiple planes
+ * usage. The KMS API was not atomic, so you'd have to update the primary
+ * plane and then the overlay planes with distinct IOCTL's. This could lead
+ * to tearing and also some trouble related to blocking, so the atomic
+ * API was proposed to fix these problems.
+ * 
+ * With the introduction of the KMS atomic API, all the planes could get
+ * updated in a single IOCTL, using drmModeAtomicCommit(). This can be
+ * either asynchronous or fully blocking.
+ *
+ * This example assumes that you are familiar with modeset-vsync. Only
+ * the differences between both files are highlighted here.
+ */
+
+#define _GNU_SOURCE
+#include <errno.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <time.h>
+#include <unistd.h>
+#include <xf86drm.h>
+#include <xf86drmMode.h>
+#include <drm_fourcc.h>
+
+struct modeset_buf;
+struct modeset_dev;
+static int modeset_find_plane(int fd, struct modeset_dev *dev);
+static int modeset_atomic_modeset(int fd, struct modeset_dev *dev);
+static int modeset_find_crtc(int fd, drmModeRes *res, drmModeConnector *conn,
+			     struct modeset_dev *dev);
+static int modeset_create_fb(int fd, struct modeset_buf *buf);
+static void modeset_destroy_fb(int fd, struct modeset_buf *buf);
+static int modeset_setup_dev(int fd, drmModeRes *res, drmModeConnector *conn,
+			     struct modeset_dev *dev);
+static int modeset_open(int *out, const char *node);
+static int modeset_prepare(int fd);
+static void modeset_draw(int fd);
+static void modeset_draw_dev(int fd, struct modeset_dev *dev);
+static void modeset_cleanup(int fd);
+
+/*
+ * A new struct is introduced: drm_object. It stores properties of certain
+ * objects (connectors, CRTC and planes) that are used in atomic modeset setup
+ * and also in atomic page-flips (all planes updated in a single IOCTL).
+ */
+
+struct drm_object {
+	drmModeObjectProperties *props;
+	drmModePropertyRes **props_info;
+};
+
+struct modeset_buf {
+	uint32_t width;
+	uint32_t height;
+	uint32_t stride;
+	uint32_t size;
+	uint32_t handle;
+	uint8_t *map;
+	uint32_t fb;
+};
+
+struct modeset_dev {
```

Comment:
I find this name confusing. I thought it referred to the device, but it turns out the code is creating one of these for each connector. So the name should be something about a connector or an output.

I was wondering why there was a list of DRM devices, but there isn't.

Reply:
Done.